### ISSUE TITLE
Micro stutter randomly occurs

### ISSUE THREAD

**Issue by Velonex76** [coelckers/prboom-plus#524]
Whenever I use prboom plus 2.6.2 and dev builds above it my game will randomly stutter and skip from frames. I am running the game on a machine that is more than capable of running even the most demanding maps. I also run the game in opengl rendering in case that may be part of the problem. I have notice this also occurs in DSDA DOOM, running the latest version, so the issue can be fundamental to prboom or just my pc.

**Comment by kraflab** [comment]
Is there any consistent trigger for when it happens? For instance, I've heard some players experience a stutter the first time certain sounds play. Is it completely random or is there a common factor at play?

**Comment by JadingTsunami** [comment]
Do you have uncappted framerate enabled?

Can you reproduce this issue with capped framerate?

**Comment by Velonex76** [comment]
> Is there any consistent trigger for when it happens? For instance, I've heard some players experience a stutter the first time certain sounds play. Is it completely random or is there a common factor at play?

Now that you mention that it is entirely possible that is the case since I remember it happening whenever a new enemy appears. And if that is the case is there anything I can do to resolve it?

**Comment by Velonex76** [comment]
> Do you have uncappted framerate enabled?
> 
> Can you reproduce this issue with capped framerate?

I have yet to try capping it with turning off uncapped frame however I have tried capping the frame rate via v sync as well as capping it with Nvidia control panel.

**Comment by JadingTsunami** [comment]
> I have yet to try capping it with turning off uncapped frame

Please try capping framerate using the in-game configuration option and see if the issue is resolved.

**Comment by JadingTsunami** [comment]
> Turning off "uncapped framerate" does resolve the issue however I do prefer the smoothness that comes having a high framerate so I would like to find a way to keep it on.

You can try vsync and see if the issue resolves. If not, you will need to play with capped framerate. You can increase the fps limit in dsda-doom to try and find a smooth fps that avoids the stutter.

**Comment by Velonex76** [comment]
> > Turning off "uncapped framerate" does resolve the issue however I do prefer the smoothness that comes having a high framerate so I would like to find a way to keep it on.
> 
> You can try vsync and see if the issue resolves. If not, you will need to play with capped framerate. You can increase the fps limit in dsda-doom to try and find a smooth fps that avoids the stutter.

ok I have been testing that out with some results, is there a recommended fps cap per hz of the monitor?

**Comment by Velonex76** [comment]
can adjusting the "texture options" fix this issue for me? The only change I make to it is turning off "texture filtering".

**Comment by vocaab** [comment]
Yeah I have been getting this too. Easiest way to reproduce it is fire up TNT and sit and monitor the frames. It will just randomly dip. Settings unlimited framerate vsync on. I have 240hz panel too so it will dip from 240-239 to 234 randomly by doing nothing.

**Comment by Velonex76** [comment]
> Yeah I have been getting this sometimes too. Easiest way to reproduce it for me to is fire up TNT m1 and sit and monitor the frames. It will just randomly dip. Even panning the camera around with the arrow keys is jerky. Settings unlimited framerate/vsync on. I have a 240hz panel too so it will dip from 240-239 to 234 randomly by doing nothing. It happens in DSDA as well, though dsda feels more stable to me. My suggestion would to disable vsync and use RTSS/or nvidia control panel and limit the frames there.

Yeah I have done that and it does reduce the issue but it still exists

**Comment by Velonex76** [comment]
so my best solution to the issue to play in software mode since it removes the stutters. Software mode play much smoother than I remember and I like it's lighting so I don't mind it.

**Comment by Velonex76** [comment]
All I can do now is hope the issue is resolved in the next update for either DSDA or PrBoom +

**Comment by rrPKrr** [comment]
I think I accidentally stumbled upon a workaround for that problem. I also get some random hitches in PrBoom+/DSDA-Doom from time to time in OpenGL mode (never happens in software mode though). They are almost negligible in my case as I play on g-sync monitor, but I guess this is exactly the problem that others have described here already.

First of all these are my settings (in DSDA-Doom):

videomode                         "OpenGL"
screen_resolution                "960x540"
use_fullscreen                     1
exclusive_fullscreen             0
render_vsync                      0
uncapped_framerate           1
dsda_fps_limit                     105

The display is set to 120Hz, g-sync mode. V-sync is enabled in nvidia control panel for DSDA-Doom.

It seems like I can consistently eliminate those random stutters by simply changing Exclusive Fullscreen to "on" and then back to "off" for the duration of the level. And apparently the game reintroduces stutters after the map change. But if I go to the options menu and change exclusive fullscreen to on and off again, then once again it will fix random hitches for the duration of that map.

**Comment by kraflab** [comment]
Does it have to be exclusive fullscreen or does it get fixed if you switch to software and back to opengl? Trying to find out if reinitializing the video mode fixes it (as opposed to specifically the fullscreen toggle).

**Comment by Velonex76** [comment]
I feel that adding something like "precache_sound" that is used in Woof would resolve the issue since it usually comes up when in sounds are activated such as enemies, getting a weapon, etc.

**Comment by rrPKrr** [comment]
> I feel that adding something like "precache_sound" that is used in Woof would resolve the issue since it usually comes up when in sounds are activated such as enemies, getting a weapon, etc.

Well, the problem is that after debugging this issue for some time the only hiccup I have managed to catch is SDL_GL_SwapWindow() function... I personally play tha game with 105 fps cap (in g-sync mode), and for 99% of time the whole program manages to finish all calculations in under ~9ms no problem, but then at random times and out of the blue, SDL_GL_SwapWindow() can take up to 45-50ms, and that's where the stutter happens. Unless there are multiple sources of the stutter of course, but once again I have noticed 0 frametime fluctuations caused by sound related code (or any other code really).

Can somebody else time SDL_GL_SwapWindow() and confirm or deny whether this the case?

**Comment by rrPKrr** [comment]
Yeah, so I added a crude function to DSDA-Doom https://github.com/rrPKrr/dsda-doom/commit/2b2b5baced08f3c7f4bdb446eabddf9812d4e0ce , and timed D_DoomLoop(), D_Display() and I_FinishUpdate() while playing E1M1 normally at 105fps (105 is chosen because it's 35x3):

![FinishUpdate](https://user-images.githubusercontent.com/110629864/202570148-e13a73e6-6d3e-4a76-857d-0ca02084789f.jpg)

As you can see from the log, D_DoomLoop() shows consistent 28-29ms (that's because I use itoa() since approximation doesn't play any role here, as those +/- 1 ms inaccuracies aren't what's causing the problem). During these ~28 ms D_Display() consistently updates every ~9.5ms 3 times and framerate sits at 105. But when the stutter occurs D_DoomLoop() tanks down to ~79ms, and the only place where it spends so much time is I_FinishUpdate(). To be more precise it sits in gld_Finish() > SDL_GL_SwapWindow() for no apparent reasaon.

I thought it was a debugging environment that might have caused these spikes, but now after adding a logging function and playing the game normally I experience exactly the same behavior. ¯\_(ツ)_/¯

**Comment by kraflab** [comment]
I guess you already implemented a timing solution, but for future reference you are looking for dsda/time. There's an example use here: https://github.com/kraflab/dsda-doom/blob/dcb92d049f30d8b7501f14d6d2fba8a68dfef3cc/prboom2/src/dsda/key_frame.c#L390. Wrapping it in a log to the console would just be a little extra 🙂

This is a really great investigation, thanks for digging in so deep! For the first time I'm feeling really optimistic that it will be solved 😄

**Comment by rrPKrr** [comment]
To be honest there is one more thing that I am not quite sure how to interpet.

It seems like there are actually two things going on simultaneously, but due to some similarities and hard to distinguish differences people just call them stutters. But here is what I noticed. I have already mentioned SDL_GL_SwapWindow(). In game it appears like your regular stutter (read microfreeze). But then sometimes this happens:

![doomloop](https://user-images.githubusercontent.com/110629864/202654086-5b5ad98f-cbc5-42e3-a7f8-4dc6c83afe08.jpg)

Because of very low frametimes it's quite easy to mistake this artifact for a hiccup, but if you look very carefully, you will notice as if it was some repeating (or broken) interpolated frame which doesn't really belong to the current camera position. What's strange about this is that I couldn't find anything that could add those extra milliseconds to D_DoomLoop() which would cause it to play 1 extra frame. So, yeah...

https://www.youtube.com/watch?v=GRWbIoIR04c

**Comment by rrPKrr** [comment]
Just a minor update, because I began chasing ghosts after doubting my own findings...

I have simply commented out SDL_GL_SwapWindow(sdl_window) and replayed a demo so that the game could play normally in the background (because commenting out SDL_GL_SwapWindow() makes the game to not show on the screen). No single frame takes longer than ~9.5ms to finish this way. Not even one. So, yeah, now I am 100% sure. 😅

**Comment by kraflab** [comment]
What if you play a demo without it commented out? Does it stutter?

**Comment by rrPKrr** [comment]
> What if you play a demo without it commented out? Does it stutter?

Yes, SDL_GL_SwapWindow() tanks regardless if's it's demo playback or a normal play.

**Comment by kraflab** [comment]
And you don't have vsync on I assume?

**Comment by rrPKrr** [comment]
> And you don't have vsync on I assume?

videomode "OpenGL"
screen_resolution "960x540"
use_fullscreen 1
exclusive_fullscreen 0
render_vsync 0
uncapped_framerate 1
dsda_fps_limit 105

The display is set to 120Hz, g-sync mode. V-sync is enabled in nvidia control panel for DSDA-Doom.

V-sync is on because it's needed for g-sync to work properly (just to clarify it works differently than your normal v-sync).

But before you ask, no, I have already checked and SDL_GL_SwapWindow() still decides to waste time randomly regardless if it's in a g-sync mode, regular fixed refresh rate v-sync mode or if all syncing methods are forced off completely.

**Comment by fabiangreffrath** [comment]
We could try different settings for this function
https://wiki.libsdl.org/SDL_GL_SetSwapInterval

**Comment by rrPKrr** [comment]
🤔 https://gamedev.stackexchange.com/questions/181327/are-there-any-good-techniques-for-reducing-or-smoothing-stutter-after-a-longer-f

**Comment by rrPKrr** [comment]
> We could try different settings for this function https://wiki.libsdl.org/SDL_GL_SetSwapInterval

It's controlled by the in-game v-sync option, no? But other than that it can also be set to -1. I have already tried that.

**Comment by kraflab** [comment]
Correct me if I'm wrong, but the rendering may be asynchronous to some extent, right? As in, there may be buffered draw commands that are only guaranteed to be done when SDL_GL_SwapWindow returns? If that's the case, then you wouldn't actually know *what* is slow on that frame, even though the slowdown shows up at the swap.

**Comment by rrPKrr** [comment]
Wait a second... After reading this comment https://gamedev.stackexchange.com/questions/181327/are-there-any-good-techniques-for-reducing-or-smoothing-stutter-after-a-longer-f I got curious so I made a completely empty loop using glfw+glew. Just glClear(GL_COLOR_BUFFER_BIT); followed by glfwSwapBuffers(window);

That actually does exactly the same thing as DSDA-Doom. Even the frametime spikes frequence looks the same.

**Comment by JadingTsunami** [comment]
Did you try running the process with the highest scheduler priority?

It's possible the OS isn't scheduling the app every time.

Else you might try `DwmFlush` after every swap buffers call.

**Comment by rrPKrr** [comment]
> Did you try running the process with the highest scheduler priority?
> 
> It's possible the OS isn't scheduling the app every time.
> 
> Else you might try `DwmFlush` after every swap buffers call.

1. Well, I tried now after you mentioned that. But nah, it does nothing.
2. DwmFlush did a lot of things. First of all it reduced fps from capped 105 to variable 89-94 for some reason. Then it changed SDL_GL_SwapWindow() times so that now they matched times spent in D_Display():
![dwmflush](https://user-images.githubusercontent.com/110629864/202865194-9b4707a0-a9cb-4340-a0de-9d6b70cf869e.jpg)
But, as you can see, at some point SDL_GL_SwapWindow() times still skyrocketed for no reason. So, sadly that made things only worse.

**Comment by JadingTsunami** [comment]
>DwmFlush did a lot of things. 

Oh, interesting. Could you perhaps try enabling multisampling? I am interested to see what effect this would have.

**Comment by rrPKrr** [comment]
> Oh, interesting. Could you perhaps try enabling multisampling? I am interested to see what effect this would have.

Performance wise it's the same. VIsually I don't see any difference either. Why? What should've theorethically happened?

**Comment by JadingTsunami** [comment]
> Performance wise it's the same. VIsually I don't see any difference either. 

Hmm, perhaps you'll have to try enabling it manually then. You could try in your test program.

`glfwWindowHint(GLFW_SAMPLES, 4);`

You *probably* should not need to enable it, but in case you can try:

`glEnable(GL_MULTISAMPLE);`

It should not hurt either way.

> Why? What should've theorethically happened?

I am suspicious that some interaction between compositioning and the buffer swap is happening. By enabling multisampling I *think* this will avoid some possible routes that could stall. But I am not totally sure here. I don't have any hardware new enough to test this out so I am having to guess a bit from here.

You should have seen at least some visual change so I am thinking you can try the above call.

**Comment by rrPKrr** [comment]
Nope, nothing. It doesn't affect frametimes with or without glEnable(GL_MULTISAMPLE).

One other thing I have tried in a blank program is setting GLFW_CONTEXT_VERSION_MAJOR and GLFW_CONTEXT_VERSION_MINOR to 4.6 (as far as I understand it DSDA uses 2.1), but it doesn't do anything either.

I wonder how does Doom (2016) swap buffers in an OpenGL mode. I've downloaded it, but frametimes are absolutely perfect no matter how long you play. There must be something. -.-

I'll see if I can write an empty project in OpenGL without glfw/sdl/etc, and see how swapping buffers would work there tomorrow.

**Comment by JadingTsunami** [comment]
>I wonder how does Doom (2016) swap buffers in an OpenGL mode.

But are you in windowed mode there? I think the problem is non-exclusive fullscreen which suggests a compositing issue.

**Comment by rrPKrr** [comment]
> But are you in windowed mode there? I think the problem is non-exclusive fullscreen which suggests a compositing issue.

Both exclusive fullscreen and borderless fullscreen modes work flawlessly in Doom 2016.

I have tried creating an empty opengl project without sdl2/glew... That said I noticed that glFinish() called right after swapping buffers offloads all the work out of buffer swapping function completely. So, maybe glFinish() could be called in a parralel thread? 🤔

**Comment by kraflab** [comment]
Is it possible the applications that don't have this issue for you are using triple buffering? Might be a dumb question, I don't know much about the details :^)

**Comment by rrPKrr** [comment]
> Is it possible the applications that don't have this issue for you are using triple buffering? Might be a dumb question, I don't know much about the details :^)

As far as I know the only way to enable tripple buffering is to turn it on via driver options, and there is no way to enable it programmatically. The tripple buffering is forced off globally, so it's very unlikely. That said tripple buffering was one of the first things I tried to use in dsda doom to see if it'd make any difference (might've been a stupid thing to try :D).

Sadly no, the solution is not that simple. But feel free to suggest whatever else will come up to your mind. I will try it anyway (if I haven't tested that already).

**Comment by rrPKrr** [comment]
I was partially wrong about the game stuttering in a non v-sync mode. It doesn't. The catch is that you need to disable any framerate cap in conjunction with disabled v-sync no matter how high the cap value is. Even if it's set to 1000 the fps will tank down to ~20.

V-sync off + fps cap off = stutter free. 

That could've been all fine and dandy, but then we get screen tearing. It seems like setting "Maximum pre-rendered frames" to very high values (for example 8) significantly improves the game smoothness. At least on a g-sync monitor. If somebody will stumble upon this comment in the future, give it a try.

**Comment by rrPKrr** [comment]
I have managed to find a "fix", finally.

There isn't much info on that, but according to this page: https://learn.microsoft.com/en-us/gaming/game-bar/known-issues

> Depending on video card driver implementation a game running in any kind of full-screen mode will potentially be rendered without Desktop Window Manager (DWM) involvement

It seems like there is a bug in Nvidia drivers and/or Windows that causes OpenGL buffer swapping to not play nicely with Desktop Window Manager. There is a hackish way to make OpenGL apps to work along DWM though.

First you need to enable Xbox Game Bar in Windows settings:
![Untitled2](https://user-images.githubusercontent.com/110629864/206431114-2177f250-d1c3-4c85-b585-adb5d57d56fc.jpg)

After that while in game press Win+G, and pin ANY element of the overlay (they won't be shown in game, as according to the same page the overlay doesn't work properly in OpenGL/Vulkan apps in fullscreen modes):
![Untitled](https://user-images.githubusercontent.com/110629864/206432233-8361656a-850b-42c4-9e3f-819a7d43549e.jpg)

Done. The problem is magically fixed. Buffer Swapping in DSDA-Doom doesn't take longer than nanoseconds. Ever. The stuttering is gone in v-sync/g-sync modes.
